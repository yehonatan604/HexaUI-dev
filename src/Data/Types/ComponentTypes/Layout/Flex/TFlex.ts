import { TComponent, ComponentOptions } from "../../TComponent";

export type TFlex = TComponent<"div"> & {
    options?: FlexOptions;
}

export type FlexOptions = ComponentOptions & {
    direction?: string;
    justify?: string;
    align?: string;
    wrap?: string;
}