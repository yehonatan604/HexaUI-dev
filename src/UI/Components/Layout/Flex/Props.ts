import { TComponent } from "../../../../Data/Types/ComponentTypes/TComponent";

export type FlexProps = TComponent<"div"> & {
    direction?: string;
    justify?: string;
    align?: string;
    wrap?: string;
};