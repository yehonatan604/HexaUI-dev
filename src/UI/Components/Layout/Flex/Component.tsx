import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import { FlexWrap } from "../../../../Data/Constants/FlexWrap";
import { TFlex } from "../../../../Data/Types/ComponentTypes/Layout/Flex/TFlex";

const Flex = (props: TFlex) => {
  const { children, options, className, ...componentProps } = props;

  // Options
  const directionClass = options?.direction || FlexDir.Row;
  const justifyClass = options?.justify || FlexTypes.Start;
  const alignClass = options?.align || FlexTypes.Start;
  const wrapClass = options?.wrap || FlexWrap.NoWrap;

  // JSX
  return (
    <div
      className={`flex flex-${directionClass} justify-${justifyClass} items-${alignClass} flex-${wrapClass} ${className}`}
      {...componentProps}
    >
      {children}
    </div>
  );
};

export default Flex;
