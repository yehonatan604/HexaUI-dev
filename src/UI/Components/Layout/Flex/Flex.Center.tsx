import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import { FlexWrap } from "../../../../Data/Constants/FlexWrap";
import { TFlex } from "../../../../Data/Types/ComponentTypes/Layout/Flex/TFlex";
import Flex from "./Flex";

const FlexCenter = (props: TFlex) => {
  const { children, options, className, ...componentProps } = props;

  // Options
  const direction = options?.direction || FlexDir.Row;
  const wrap = options?.wrap || FlexWrap.NoWrap;

  // JSX
  return (
    <Flex
      options={{ ...options, justify: FlexTypes.Center, align: FlexTypes.Center }}
      className={`flex-${direction} flex-${wrap} ${className}`}
      {...componentProps}
    >
      {children}
    </Flex>
  );
};

export default FlexCenter;
