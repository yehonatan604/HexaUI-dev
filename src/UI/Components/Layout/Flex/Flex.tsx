import { useEffect } from "react";
import { FlexDir } from "../../../../Data/Constants/FlexDirection";
import { FlexTypes } from "../../../../Data/Constants/FlexTypes";
import { FlexWrap } from "../../../../Data/Constants/FlexWrap";
import { TFlex } from "./types/TFlex";

const Flex = (props: TFlex) => {
  const { children, options, className, center, end, ...componentProps } = props;

  // Options
  const bgVariant = options?.bgVariant || "";
  const textVariant = options?.textVariant || "";
  const directionClass = options?.direction || FlexDir.Row;
  const justifyClass = center
    ? FlexTypes.Center
    : end
    ? FlexTypes.End
    : options?.justify || FlexTypes.Start;
  const alignClass = center
    ? FlexTypes.Center
    : end
    ? FlexTypes.End
    : options?.align || FlexTypes.Start;
  const wrapClass = center ? FlexWrap.Wrap : options?.wrap || FlexWrap.NoWrap;

  useEffect(() => {
    if (center && end) {
      console.error("Flex component can't have both center and end props");
    }
  }, [center, end]);

  // JSX
  return (
    <div
      className={`flex flex-${directionClass} bg-${bgVariant} text-${textVariant} justify-${justifyClass} items-${alignClass} flex-${wrapClass} ${className}`}
      {...componentProps}
    >
      {children}
    </div>
  );
};

export default Flex;
