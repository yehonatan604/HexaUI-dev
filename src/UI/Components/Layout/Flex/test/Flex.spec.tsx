import { render, screen } from "@testing-library/react";
import Flex from "../Flex"; // Adjust the import path
import { describe, it, expect, vi } from "vitest";

describe("Flex Component", () => {
  it("should render children correctly", () => {
    render(
      <Flex>
        <div>Child 1</div>
        <div>Child 2</div>
      </Flex>
    );

    // Check if both children are rendered
    expect(screen.getByText("Child 1")).toBeInTheDocument();
    expect(screen.getByText("Child 2")).toBeInTheDocument();
  });

  it("should apply default flex direction and alignment", () => {
    render(
      <Flex>
        <div>Child 1</div>
      </Flex>
    );

    const flexElement = screen.getByText("Child 1").parentElement;

    // Check if the default flex direction is 'row' and default alignment is applied
    expect(flexElement).toHaveClass(
      "flex",
      "flex-row",
      "justify-start",
      "items-start",
      "flex-nowrap"
    );
  });

  it("should apply custom flex direction and alignment from options", () => {
    render(
      <Flex
        options={{
          direction: "col",
          justify: "end",
          align: "center",
          wrap: "wrap",
        }}
      >
        <div>Child 1</div>
      </Flex>
    );

    const flexElement = screen.getByText("Child 1").parentElement;

    // Check if the custom flex direction, justify, align, and wrap classes are applied
    expect(flexElement).toHaveClass(
      "flex",
      "flex-col",
      "justify-end",
      "items-center",
      "flex-wrap"
    );
  });

  it("should apply center prop correctly", () => {
    render(
      <Flex center>
        <div>Centered Child</div>
      </Flex>
    );

    const flexElement = screen.getByText("Centered Child").parentElement;

    // Check if both justify and align are centered
    expect(flexElement).toHaveClass(
      "flex",
      "flex-row",
      "justify-center",
      "items-center",
      "flex-wrap"
    );
  });

  it("should apply end prop correctly", () => {
    render(
      <Flex end>
        <div>End Aligned Child</div>
      </Flex>
    );

    const flexElement = screen.getByText("End Aligned Child").parentElement;

    // Check if both justify and align are set to end
    expect(flexElement).toHaveClass(
      "flex",
      "flex-row",
      "justify-end",
      "items-end",
      "flex-nowrap"
    );
  });

  it("should log an error when both center and end props are provided", () => {
    const consoleErrorSpy = vi.spyOn(console, "error");

    render(
      <Flex center end>
        <div>Error Test</div>
      </Flex>
    );

    // Expect the error to be logged
    expect(consoleErrorSpy).toHaveBeenCalledWith(
      "Flex component can't have both center and end props"
    );

    consoleErrorSpy.mockRestore(); // Clean up the spy
  });

  it("should render and match the snapshot", () => {
    const { asFragment } = render(
      <Flex
        options={{
          direction: "col",
          justify: "center",
          align: "end",
          wrap: "wrap",
        }}
      >
        <div>Snapshot Test</div>
      </Flex>
    );

    // Snapshot the Flex component structure
    expect(asFragment()).toMatchSnapshot();
  });
});
