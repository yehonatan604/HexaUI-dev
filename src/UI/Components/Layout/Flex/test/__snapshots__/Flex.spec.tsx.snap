// Vitest Snapshot v1

exports[`Flex Component > should render and match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col bg- text- justify-center items-end flex-wrap undefined"
  >
    <div>
      Snapshot Test
    </div>
  </div>
</DocumentFragment>
`;
